components/camera/rtsp/rtsp.go

```diff
@@ -0,0 +1,241 @@
+					rc.logger.Warnw("The rtsp server responded with non-OK status", "url", rc.u, "status code", res.StatusCode)
+					badState = true
+				}
+				if badState {
+					if err = rc.reconnectClient(); err != nil {
+						rc.logger.Warnw("cannot reconnect to rtsp server", "error", err)
+					} else {
+						rc.logger.Infow("reconnected to rtsp server", "url", rc.u)
+					}
+				}
+			} else {
+				return
+			}
+		}
+	}, rc.activeBackgroundWorkers.Done)
+}
+
+// reconnectClient reconnects the RTSP client to the streaming server by closing the old one and starting a new one.
+func (rc *rtspCamera) reconnectClient() (err error) {
+	if rc == nil {
+		return errors.New("rtspCamera is nil")
+	}
+	if rc.client != nil {
+		err := rc.client.Close()
+		if err != nil {
+			rc.logger.Debugw("error while closing rtsp client:", "error", err)
+		}
+	}
+	// replace the client with a new one, but close it if setup is not successful
+	client := &gortsplib.Client{}
+	rc.client = client
+	var clientSuccessful bool
+	defer func() {
+		if !clientSuccessful {
+			if errClose := rc.client.Close(); errClose != nil {
+				err = multierr.Combine(err, errClose)
+			}
+		}
+	}()
+	err = rc.client.Start(rc.u.Scheme, rc.u.Host)
+	if err != nil {
+		return err
+	}
+	mjpegFormat, mjpegDecoder := mjpegDecoding()
+
+	tracks, baseURL, _, err := rc.client.Describe(rc.u)
+	if err != nil {
+		return err
+	}
+	track := tracks.FindFormat(&mjpegFormat)
+	if track == nil {
+		return errors.New("MJPEG track not found")
+	}
+	_, err = rc.client.Setup(track, baseURL, 0, 0)
+	if err != nil {
+		return err
+	}
+	// On packet retreival, turn it into an image, and store it in shared memory
+	rc.client.OnPacketRTP(track, mjpegFormat, func(pkt *rtp.Packet) {
+		img, err := mjpegDecoder(pkt)
+		if err != nil {
+			return
+		}
+		if img == nil {
+			return
+		}
+		rc.latestFrame.Store(img)
+		if !rc.gotFirstFrameOnce {
+			rc.gotFirstFrameOnce = true
+			close(rc.gotFirstFrame)
+		}
+	})
+	_, err = rc.client.Play(nil)
+	if err != nil {
+		return err
+	}
+	clientSuccessful = true
+	return nil
+}
+
+// NewRTSPCamera creates a camera client using RTSP given the server URL.
+// Right now, only supports servers that have MJPEG video tracks.
+func NewRTSPCamera(ctx context.Context, attrs *Attrs, logger golog.Logger) (camera.Camera, error) {
+	u, err := url.Parse(attrs.Address)
+	if err != nil {
+		return nil, err
+	}
+	gotFirstFrame := make(chan struct{})
+	rtspCam := &rtspCamera{
+		u:             u,
+		logger:        logger,
+		gotFirstFrame: gotFirstFrame,
+	}
+	err = rtspCam.reconnectClient()
+	if err != nil {
+		return nil, err
+	}
+	cancelCtx, cancel := context.WithCancel(context.Background())
+	reader := gostream.VideoReaderFunc(func(ctx context.Context) (image.Image, func(), error) {
+		select {
+		case <-cancelCtx.Done():
+			return nil, nil, cancelCtx.Err()
+		case <-ctx.Done():
+			return nil, nil, ctx.Err()
+		default:
+			<-rtspCam.gotFirstFrame // block until you get the first frame
+		}
```

Comment:
Ah so this would've had to be (no default)
```
		select {
 		case <-cancelCtx.Done():
 			return nil, nil, cancelCtx.Err()
 		case <-ctx.Done():
 			return nil, nil, ctx.Err()
 		case <-rtspCam.gotFirstFrame: // block until you get the first frame
 		}
```
to get rid of the issue. That way the select could block until one of those conditions are true. As of now it could still not see a `Done()`, fall into the default case, then block indefinitely on `<-rtspCam.gotFirstFrame` if the goroutine's parent function tried to kill it before sending a frame. Not a big deal though. Just wanted to call it out in the _very_ unlikely chance this gets triggered.

Reply:
we should probably fix this. unlikely is possible

Reply:
I can change it to a (WaitChan 10s) and then continue to look for the Done() again

Reply:
bug fix incoming

Reply:
https://github.com/viamrobotics/rdk/pull/1653#discussion_r1082883572

Eric mentioned that the block you propose has its own problems, so I need to create a way to always be listening on the contexts. Essentially the easiest way will actually just putting in the select statement twice - once without the blocking gotFrame, and one with.